### PR TITLE
separate debug symbols into a dbg file

### PR DESCRIPTION
**Pull request by sarathnandu**
### Description 
_Add a comprehensive description of proposed changes_

Fixes # - _issue number(s) if exists_

### Type of change

_Choose one or multiple, leave empty if none of the other choices apply_

_Add a respective label(s) to PR if you have permissions_

- [ ] bug fix - _change that fixes an issue_
- [x ] new feature - _change that adds functionality_
- [ ] tests - _change in tests_
- [ ] infrastructure - _change in infrastructure and CI_
- [ ] documentation - _documentation update_

### Tests

- [ ] added - _required for new features and some bug fixes_
- [x ] not needed

### Documentation

- [ ] updated in # - _add PR number_
- [ ] needs to be updated
- [ ] not needed

### Breaks backward compatibility
- [x ] Yes
- [ ] No
- [ ] Unknown

### Notify the following users
_List users with `@` to send notifications_

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -94,6 +94,7 @@ endif()
 
 option(TBB_TEST "Enable testing" ON)
 option(TBB_EXAMPLES "Enable examples" OFF)
+option(TBB_STRIP_SYMBOLS "Enable examples" OFF)
```
Comment (isaevil) [inline CMakeLists.txt:97]: 1. Copy-paste description
2. In my opinion, name of this variable doesn't reflect what it actually does. It should highlight that it strips debug info, and it should do it only on Linux.

File: src/tbb/CMakeLists.txt [hunk 2]
```diff
@@ -126,6 +126,16 @@ target_link_libraries(tbb
     ${TBB_COMMON_LINK_LIBS}
 )
 
+#strip symbols to a separate dbg file
+if(TBB_STRIP_SYMBOLS)
+    add_custom_command(TARGET tbb POST_BUILD
+        COMMAND objcopy --only-keep-debug $<TARGET_FILE:tbb> $<TARGET_FILE:tbb>.dbg
+        COMMAND objcopy --strip-debug $<TARGET_FILE:tbb>
+        COMMAND objcopy --add-gnu-debuglink=$<TARGET_FILE:tbb>.dbg $<TARGET_FILE:tbb>
+        COMMENT "Creating and associating .dbg file with tbb"
```
Comment (isaevil) [inline src/tbb/CMakeLists.txt:135]: Maybe check that `objcopy` command is available and if not then raise an error?

File: CMakeLists.txt (hunk 1 above)
Comment (isaevil) [inline CMakeLists.txt:97]: Maybe make this a Linux-dependent option?

File: src/tbb/CMakeLists.txt [hunk 3]
```diff
@@ -158,7 +168,12 @@ if (TBB_INSTALL)
                 COMPONENT devel
         )
     endif()
-
+    if(TBB_STRIP_SYMBOLS)
+        install(FILES
+                $<TARGET_FILE:tbb>.dbg
+                DESTINATION lib
```
Comment (isaevil) [inline src/tbb/CMakeLists.txt:174]: Add COMPONENT `devel`?

File: src/tbb/CMakeLists.txt [hunk 4]
```diff
@@ -126,6 +126,21 @@ target_link_libraries(tbb
     ${TBB_COMMON_LINK_LIBS}
 )
 
+#strip symbols to a separate dbg file
+if(TBB_LINUX_SEPARATE_DBG)
+    find_program(OBJCOPY_COMMAND objcopy)
+    if(NOT OBJCOPY_COMMAND)
+        message(STATUS "objcopy command not found in the system")
```
Comment (isaevil) [inline src/tbb/CMakeLists.txt:133]:
```suggestion
        message(WARNING "objcopy command not found in the system")
```

File: src/tbb/CMakeLists.txt [hunk 5]
```diff
@@ -126,6 +126,21 @@ target_link_libraries(tbb
     ${TBB_COMMON_LINK_LIBS}
 )
 
+#strip symbols to a separate dbg file
```
Comment (isaevil) [inline src/tbb/CMakeLists.txt:129]:
```suggestion
# Strip debug symbols into a separate .dbg file
```

File: src/tbb/CMakeLists.txt [hunk 6]
```diff
@@ -126,6 +126,21 @@ target_link_libraries(tbb
...
+if(TBB_LINUX_SEPARATE_DBG)
+    find_program(OBJCOPY_COMMAND objcopy)
+    if(NOT OBJCOPY_COMMAND)
+        message(STATUS "objcopy command not found in the system")
+    else()
+        add_custom_command(TARGET tbb POST_BUIL
```
Comment (isaevil) [inline src/tbb/CMakeLists.txt:135]:
```suggestion
        add_custom_command(TARGET tbb POST_BUILD
```

Comment (sarathnandu) [reply src/tbb/CMakeLists.txt]: updated, Thanks for catching this :)

File: src/tbb/CMakeLists.txt (hunk 4 above)
Comment (sarathnandu) [reply src/tbb/CMakeLists.txt]: Updated to warning.

**Review by isaevil — APPROVED** [review]
Looks good to me. @pavelkumbrasev what do you think?

File: src/tbb/CMakeLists.txt [hunk 7]
```diff
@@ -126,6 +126,25 @@ target_link_libraries(tbb
     ${TBB_COMMON_LINK_LIBS}
 )
 
+# Strip debug symbols into a separate .dbg file
+if(TBB_LINUX_SEPARATE_DBG)
+    if(NOT CMAKE_BUILD_TYPE STREQUAL "release")
```
Comment (pavelkumbrasev) [inline src/tbb/CMakeLists.txt:131]: Should we strip it for `debug` configuration?

Comment (sarathnandu) [reply src/tbb/CMakeLists.txt:131]: The debug symbols will not be stripped in any default configuration. If user set the BUILD_TYPE='debug' or relwithdebinfo and explicitly set the TBB_LINUX_SEPARATE_DBG switch, then the symbols get stripped to a different file and linked to binary. Whereas when BUILD_TYPE is 'release' the .dbg file has only the linkage information and hence creation of .dbg is avoided.

**Pull request merged**